## PROYECTO FINAL 
Las plataformas de streaming y los sellos discograficos necesitan entender que caracteristicas musicales hacen que una cancion sea popular o clasificar automaticamente nuevos lanzamientos dentro de su genero correcto para mejorar los algoritmos de recomendacion.
## Problema 
Clasificar manualmente el genero de cada nueva canción es costoso e ineficiente.

## SPOTIFY 
# Estudiantes 
* Daniel Barbosa 
* Maria Veronica Suarez



## Descripcion del conjunto de datos

Este conjunto de datos contiene informacion sobre canciones de Spotify pertenecientes a 125 generos musicales distintos, donde cada fila representa una sola cancion y cada columna representa un atributo especifico relacionado con esa pista. Junto con metadatos basicos como el nombre de la cancion, el artista y la etiqueta de genero, el conjunto de datos incluye varias caracteristicas de audio numericas extraidas por el sistema de analisis de audio de Spotify, tales como danceability (bailabilidad), energy (energia), loudness (sonoridad), speechiness (presencia de voz hablada), acousticness (acustica), instrumentalness (instrumentalidad), liveness (presencia de publico en vivo), valence (positividad musical), tempo (ritmo/PPM) y duration (duracion). Estas caracteristicas describen cuantitativamente las propiedades musicales y el estado de animo de cada pista, lo que hace que el conjunto de datos sea muy adecuado para el analisis de datos, la visualizacion y tareas de aprendizaje automatico (machine learning) como la clasificacion de generos, la agrupacion (clustering), los sistemas de recomendacion y la prediccion del estado de animo. Dado que los datos se proporcionan en formato CSV (valores separados por comas), la informacion es tabular, liviana y facil de cargar en herramientas como Python (Pandas), R, Excel o bases de datos SQL para un rapido procesamiento y modelado. 


## Descripcion de las columnas
* Track_id :Un identificador unico asignado a cada cancion por Spotify.

* Artists :Nombre del artista o artistas que interpretaron la cancion. Los artistas multiples estan separados por un punto y coma (;).

* Album_name :El nombre del album en el que aparece la cancion.

* Track_name :El nombre de la cancion.

* Popularity :Un valor entre 0 y 100 que indica que tan popular es la cancion. Un valor mas alto significa que es mas popular. Se basa principalmente en el numero de reproducciones y en que tan recientes son.

* Duration_ms :La duracion de la cancion en milisegundos.

* Explicit :Indica si la cancion contiene letras explicitas para adultos
    * true = Si
    * false = No o desconocido

* Danceability :Que tan adecuada es la cancion para bailar 
    * 0 = menos bailable
    * 1 = mas bailable.

* Energy :Que tan energica se siente la cancion 
    * 0 = tranquila/suave
    * 1 = muy energica/ruidosa.

* Key :La tonalidad musical de la cancion (p. ej., Do, Re, Mi). 
    * Si no se conoce, el valor es : 1.

* Loudness :La sonoridad general de la cancion en decibelios (dB).

* Mode :Indica la modalidad de la cancion .
    * 1 = Mayor , generalmente suena alegre 
    * 0 = Menor, generalmente suena triste.

* Speechiness :Mide la cantidad de contenido hablado que hay en la cancion. 
    * Cercano a 1 = Principalmente palabras habladas (p. ej., un podcast)
    * Entre 0,33 y 0,66 = Mezcla de voz y musica (p. ej., rap)
    * Por debajo de 0,33 = Principalmente musica

* Acousticness :Probabilidad de que la cancion sea acustica (1 = muy probable que sea acustica).

* Instrumentalness :Probabilidad de que la cancion no tenga voz (1 = sin canto).

* Liveness :Probabilidad de que la cancion haya sido grabada en vivo (un valor mas alto = mayor probabilidad de ser en vivo).

* Valence :Describe el estado de animo de la cancion
    * Valor alto = Alegre/positivo,
    * Valor bajo = Triste/negativo

* Tempo :La velocidad de la cancion medida en pulsaciones por minuto (PPM / BPM).

* Time_signature :Numero de tiempos por compas (generalmente entre 3 y 7).

* Track_genre :La categoria de genero musical de la cancion.






In [1]:
import os
import kagglehub
import pandas as pd

# 1. Descargamos el dataset desde Kaggle
folder_path = kagglehub.dataset_download("yashdev01/spotify-tracks-dataset")

# 2. Buscamos el archivo CSV
csv_name = [f for f in os.listdir(folder_path) if f.endswith(".csv")][0]
file_path = os.path.join(folder_path, csv_name)

# 3. Leemos el DataFrame
df = pd.read_csv(file_path)

# 4. Creamos el directorio si no existe y guardar la copia
output_dir = "../data/raw"
os.makedirs(output_dir, exist_ok=True)  # <-- Se crea la carpeta automáticamente

output_file = os.path.join(output_dir, "spotify-tracks-dataset.csv")
df.to_csv(output_file, index=False)

print("¡Archivo guardado exitosamente en:", output_file)
df.head()

¡Archivo guardado exitosamente en: ../data/raw\spotify-tracks-dataset.csv


,Unnamed: 0.1,Unnamed: 0,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,...,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,0,0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,...,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,1,1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,...,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,2,2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,...,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,3,3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,...,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,4,4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,...,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


In [2]:
df = df.drop(columns=['Unnamed: 0.1', 'Unnamed: 0'])
df.head()

,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,1,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,1,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,0,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,0,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,2,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


In [3]:
df.shape

(114000, 20)

Se tiene un dataset de 114000 filas y 20 columnas 

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   track_id          114000 non-null  str    
 1   artists           113999 non-null  str    
 2   album_name        113999 non-null  str    
 3   track_name        113999 non-null  str    
 4   popularity        114000 non-null  int64  
 5   duration_ms       114000 non-null  int64  
 6   explicit          114000 non-null  bool   
 7   danceability      114000 non-null  float64
 8   energy            114000 non-null  float64
 9   key               114000 non-null  int64  
 10  loudness          114000 non-null  float64
 11  mode              114000 non-null  int64  
 12  speechiness       114000 non-null  float64
 13  acousticness      114000 non-null  float64
 14  instrumentalness  114000 non-null  float64
 15  liveness          114000 non-null  float64
 16  valence           114000 non-nu

El conjunto de datos esta compuesto por 114.000 filas y 20 columnas. En cuanto a los tipos de variables, incluye 16 numericas, 5 categoricas y 1 booleana.

In [5]:
df.isnull().sum().sort_values(ascending=False)

artists             1
album_name          1
track_name          1
track_id            0
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
time_signature      0
track_genre         0
dtype: int64

Con este codigo se evalua los datos faltantes por columna, se observa que el conjunto de datos esta practicamente completo. Unicamente las variables: artists, album_name y track_name contienen 1 valor nulo cada una (correspondiente al mismo registro), mientras que las 19 variables restantes no presentan ningun dato ausente.


In [6]:
df.duplicated().sum()

np.int64(450)

Existen 450 filas completamente duplicadas en todo el DataFrame

In [7]:
df = df.dropna()

df = df.drop_duplicates()

df = df.reset_index(drop=True)

df

,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,1,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.7150,87.917,4,acoustic
1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,1,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.2670,77.489,4,acoustic
2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,0,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.1200,76.332,4,acoustic
3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,0,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.1430,181.740,3,acoustic
4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,2,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.1670,119.949,4,acoustic
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
113544,2C3TZjDRiAzdyViavDJ217,Rainy Lullaby,#mindfulness - Soft Rain for Mindful Meditatio...,Sleep My Little Boy,21,384999,False,0.172,0.2350,5,-16.393,1,0.0422,0.6400,0.928000,0.0863,0.0339,125.995,5,world-music
113545,1hIz5L4IB9hN3WRYPOCGPw,Rainy Lullaby,#mindfulness - Soft Rain for Mindful Meditatio...,Water Into Light,22,385000,False,0.174,0.1170,0,-18.318,0,0.0401,0.9940,0.976000,0.1050,0.0350,85.239,4,world-music
113546,6x8ZfSoqDjuNa5SVP5QjvX,Cesária Evora,Best Of,Miss Perfumado,22,271466,False,0.629,0.3290,0,-10.895,0,0.0420,0.8670,0.000000,0.0839,0.7430,132.378,4,world-music
113547,2e6sXL2bYv4bSz6VTdnfLs,Michael W. Smith,Change Your World,Friends,41,283893,False,0.587,0.5060,7,-10.889,1,0.0297,0.3810,0.000000,0.2700,0.4130,135.960,4,world-music


Se eliminan la fila con los valores nulos, se eliminan las 450 filas duplicadas y se restablece el dataframe 

In [8]:
df.shape

(113549, 20)

Se tiene un dataset de 113549 filas y 20 columnas 